test/imjoyTest.js

```diff
@@ -0,0 +1,125 @@
+import test from 'tape-catch'
+import axios from 'axios'
+
+import itkreadImageArrayBuffer from 'itk/readImageArrayBuffer'
+import vtkITKHelper from 'vtk.js/Sources/Common/DataModel/ITKHelper'
+import testUtils from 'vtk.js/Sources/Testing/testUtils'
+
+import createViewer from '../src/createViewer'
+import UserInterface from '../src/UserInterface'
+
+const testImage3DPath = 'base/test/data/input/HeadMRVolume.nrrd'
+const testImage3DPath2 = 'base/test/data/input/mri3D.nrrd'
+
+import createViewerBaseline from './data/baseline/createViewer.png'
+import createViewerSetImageBaseline from './data/baseline/createViewerSetImage.png'
+
+import * as imjoyCore from 'imjoy-core'
+import ndarray from 'ndarray'
+
+const TEST_STYLE_CONTAINER = {
+  position: 'relative',
+  width: '600px',
+  height: '600px',
+  minHeight: '600px',
+  minWidth: '600px',
+  maxHeight: '600px',
+  maxWidth: '600px',
+  margin: '0',
+  padding: '0',
+  top: '0',
+  left: '0',
+  overflow: 'hidden',
+}
+const TEST_VIEWER_STYLE = {
+  backgroundColor: [1, 1, 1],
+  containerStyle: TEST_STYLE_CONTAINER,
+}
+function applyStyle(el, style) {
+  Object.keys(style).forEach(key => {
+    el.style[key] = style[key]
+  })
+}
+
+test('Test ImJoy Plugin', async t => {
+  const gc = testUtils.createGarbageCollector(t)
+
+  const container = document.querySelector('body')
+  const viewerContainer = gc.registerDOMElement(document.createElement('div'))
+  container.appendChild(viewerContainer)
+
+  const response = await axios.get(testImage3DPath, {
+    responseType: 'arraybuffer',
+  })
+  const { image: itkImage, webWorker } = await itkreadImageArrayBuffer(
+    null,
+    response.data,
+    'data.nrrd'
+  )
+  webWorker.terminate()
+
+  const array = ndarray(itkImage.data, itkImage.size.slice().reverse())
+  const vtkImage = vtkITKHelper.convertItkToVtkImage(itkImage)
+
+  const imjoy_api = {
+    showMessage(plugin, info, duration) {
+      console.log(info)
+    },
+  }
+  const imjoy = new imjoyCore.ImJoy({
+    imjoy_api,
+  })
+  imjoy.event_bus.on('show_message', console.log)
+  imjoy.event_bus.on('add_window', async w => {
+    viewerContainer.id = w.window_id
+    applyStyle(viewerContainer, TEST_STYLE_CONTAINER)
+  })
+  await imjoy.start({ workspace: 'default' })
+  console.log('ImJoy started')
+
+  const plugin = await imjoy.pm.reloadPluginRecursively({
+    uri: 'http://localhost:9876/base/dist/index.html',
+  })
+  console.log(plugin)
+  await plugin.api.run({ config: {}, data: {} })
+  console.log(plugin.api.addImage)
```

Comment:
@oeway I am trying to set up a testing harness here.

`plugin.api.addImage` is currently `undefined` -- what am I doing wrong?

To reproduce:

```
# Checkout the branch
npm ci
npm run build
npm run test:debug
# Click the "Debug" button in the browser window
# Open the console
# Refresh the page
```

Reply:
I am on it.

Reply:
Ok, I think the issue is the plugin object you get is not the actual window instance, but a proxy plugin to the window plugin. To connect to the actual window, we need to use `createWindow`. This behavior might need to be improved, it will be more intuitive to just return the window plugin.
```js
  const proxyPlugin = await imjoy.pm.reloadPluginRecursively({
    uri: 'http://localhost:9876/base/dist/index.html',
  })
  console.log(proxyPlugin)
  // instantiate a viewer, the returned object is the api directly
  const viewer = await imjoy.pm.createWindow(null, {type: "itk-vtk-viewer"})
  console.log(viewer.addImage)
```

You can also do the following (recommended), which will give you directly the viewer object.
```js
  // const proxyPlugin = await imjoy.pm.reloadPluginRecursively({
  //   uri: 'http://localhost:9876/base/dist/index.html',
  // })
  const viewer = await imjoy.pm.createWindow(null, {src: 'http://localhost:9876/base/dist/index.html'})
  console.log(viewer.addImage)

```

Reply:
@oeway verified! Thank you so much! :1st_place_medal: :star2: